GoogleUtilities/ISASwizzler/GULObjectSwizzler.m

```diff
@@ -150,8 +150,17 @@ - (void)swizzle {
   }
 }
 
-- (void)dealloc {
-  objc_disposeClassPair(_generatedClass);
+- (void)swizzledObjectHasBeenDeallocatedWithGeneratedSubclass:(BOOL)isInstanceOfGeneratedSubclass {
+  // If the swizzled object had a different class, it most likely indicates that the object was
+  // ISA swizzled one more time. In this case it is not safe to dispose the generated class. We
```

Comment:
This seems to be dangerous. We should not swizzle a class multiple times as there might be a chance that a subclass of a generated class in a middle of a chain got disposed. We should not dispose a class when it has been created at run-time. Can we get how many times a delegate is created and swizzled? It seems to me that the number of delegate instances is small relatively to the number of NSURLSessions through out the lifetime of an app. Therefore, if we can detect that a class has been swizzled once by this framework, it shouldn't be swizzle again.

Reply:
@baolocdo Thank you for your feedback!
Unfortunately, the lifetime of the delegate is not always under our control (e.g. network requests sent by the host application). In Firebase we don't swizzle the same object two or more times, this is more about the host application or another library doing swizzling. 

As for a single generated subclass for a class - we've had some discussion on it [here](https://groups.google.com/a/google.com/d/msg/appledev-discuss/vtQkgbxK9Vk/Pey3kRb4CAAJ). Briefly, it is a good thing to do, but it requires significant changes that require cautious testing, so it may be implemented in the long term perspective. 

The current fix mitigates the issue for most (but not all) cases without introducing an undesirable memory impact, so it is not ideal, but it is one of the safest way to fix the issue in the short term perspective. This approach still looks much safer to me than the original unconditional subclass disposing (unless I am missing something).

Do you think it will make sense to publish this to fix the crash for the impacted users now but keep  work on a better fix that will cover all edge cases?

Reply:
I understand that we don't have control over the delegate instances and they can come and go. However, if Firebase framework already swizzles a class, we should not swizzle again. It's dangerous because we might not be the only SDK that swizzles this instance. There might be other SDKs in the same app that swizzles the same instance as well. Disposing one class can cause disruption of the chain.

Reply:
@baolocdo Right, there are some cases not covered by the fix. Would you be able to suggest a better option?

Reply:
There are 2 things I am trying to suggest here:
1. Check if the class has been swizzled out by this framework. If it has been swizzled, do not do it again.
2. The class should not be disposed once created. This can cause unexpected behavior.

Reply:
As discussed offline, we will consider the this fix a a long term solution for the issue and merge this PR as the short term fix to mitigate the existing crash.